Source/DafnyRuntime/DafnyRuntime.go

```diff
@@ -692,183 +692,412 @@ func (seq Seq) VerbatimString(asLiteral bool) string {
-  contents []interface{} // stored as a flat one-dimensional slice
-  dims     []int
-}
-
-func newArray(dims ...Int) *Array {
+// Array is the general interface for arrays. Conceptually, it contains some
+// underlying storage (a slice) of elements, together with a record of the length
+// of each dimension. Thus, this general interface supports 1- and multi-dimensional
+// Dafny arrays.
+//
+// All array indices are given as the Go type "int". This is what the Dafny implementation
+// refers to as "target-language array-index type". However, the dimension lengths
+// are given as "big.Int".
+
+type Array interface {
+  dimensionCount() int
+  dimensionLength(dim int) int
+  ArrayGet1(index int) interface{}
+  ArraySet1(value interface{}, index int)
+  anySlice(lo, hi Int) []interface{}
+  // specializations
+  ArrayGet1Byte(index int) byte
+  ArraySet1Byte(value byte, index int)
+  ArrayGet1Char(index int) Char
+  ArraySet1Char(value Char, index int)
+}
+
+/***** newArray *****/
+
+// Multiply the numbers in "dims" and return the product as an "int".
+// If the produce doesn't fit in an "int", panic with the message that the
+// array-size limit has been exceeded.
+// It is expected that len(dims) is at least 1 and that each number in
+// dims is non-negative.
+func computeTotalArrayLength(dims ...Int) int {
+  product := dims[0]
+  for i := 1; i < len(dims); i++ {
+    product = product.Times(dims[i])
+  }
+  if IntOf(0x8000_0000).Cmp(product) <= 0 {
+    panic(fmt.Sprintf("array size exceeds memory limit: %v", String(product)))
+  }
+  totalLength := product.Int()
+  return totalLength
+}
+
+// NewArrayFromExample returns a new Array.
+// If "init" is non-nil, it is used to initialize all elements of the array.
+// "example" is used only to figure out the right kind of Array to return.
+// If "init" is non-nil, the types of "example" and "init" must agree.
+func NewArrayFromExample(example interface{}, init interface{}, dims ...Int) Array {
+  numberOfDimensions := len(dims)
   intDims := make([]int, len(dims))
-  size := 1
-  for d := len(dims) - 1; d >= 0; d-- {
-    //    sizes[d] = size
+  totalLength := computeTotalArrayLength(dims...)
+  // If the previous line does not panic, then the .Int() conversions in
+  // the following loop will succeed.
+  for d := 0; d < numberOfDimensions; d++ {
     intDims[d] = dims[d].Int()
-    size *= intDims[d]
   }
-  // Bypass the SeqOf constructor to avoid defensive copy
-  contents := make([]interface{}, size)
-  return &Array{
-    contents: contents,
+
+  if totalLength == 0 {
+    return newZeroLengthArray(intDims)
+  }
+
+  // Inspect the type of "example" to consider Array specialization
+  if _, ok := example.(byte); ok {
+    arr := make([]byte, totalLength)
+    if init != nil {
+      x := init.(byte)
+      for i := range arr {
+        arr[i] = x
+      }
+    }
+    return &ArrayForByte{
+      contents: arr,
+      dims:     intDims,
+    }
+  }
+  if _, ok := example.(Char); ok {
```

Comment:
It seems that `DafnyRuntime.go` defines both `Char` and `CodePoint` as type aliases for `rune`:

``` go
type Char rune
type CodePoint rune
```

Doesn't this mean that the specialization for `Char` is really a specialization for `rune`, and therefore the array optimizations already work for both `Char` and `CodePoint`?

Reply:
Nope, a [type declaration](https://go.dev/ref/spec#Type_definitions) like those in Go creates a new distinct type, so it's like `newtype` in Dafny. It means on this line you're specifically checking for `Char`, and the type test will not succeed if given a `CodePoint` or even a plain `rune`. I confirmed this with a quick test:

```go
func TestCodePointCheck(t *testing.T) {
	var cp interface{} = CodePoint('D')
	_, ok := cp.(Char)
	if (!ok) {
		t.Error("Oh shoot: ", cp)
	}
}
```

```
=== RUN   TestCodePointCheck
    DafnyRuntime_test.go:16: Oh shoot:  'D'
--- FAIL: TestCodePointCheck (0.00s)
FAIL
exit status 1
```

Note that the value prints as `'D'` rather than `D` precisely because it's a different type that has a `String()` method that adds the single quotes, which is the whole reason I introduced `CodePoint`: so it's a distinct type that behaves differently.

Reply:
Thanks. I have fixed it.